### ISSUE TITLE
Travis CI appears to be broken

### ISSUE THREAD

**Issue by mgedmin** [buildout/buildout#487]
The last successful build was 2 months ago: https://travis-ci.org/buildout/buildout/builds/553658470

Today I merged a typo fix in a doctest and the build is failing: https://travis-ci.org/buildout/buildout/builds/574320041

The Python 2.7 build fails three tests, all Python 3.x builds hang.

**Comment by mgedmin** [comment]
I've enabled weekly cron builds on Travis CI so that in the future we'll learn about this kind of breakage sooner.

**Comment by icemac** [comment]
On branch https://github.com/buildout/buildout/tree/fix-487 I switched TravisCI to use tox.
See https://travis-ci.org/buildout/buildout/builds/574705099

This does not help with the hanging Python 3 builds, so I requested the TravisCI support to activate the debug mode so it is possible to ssh into the CI system.

Locally I detected that the hanging test takes about 3 seconds on Python 2.7 bit nearly 80 seconds on Python 3.4 and 3.5.

The tox environment for Python 3.6 cannot be installed on my machine because of `TypeError: can't pickle zipimport.zipimporter objects` while installing `bobo`.

**Comment by mgedmin** [comment]
Another interesting thing: 2 months ago all Travis jobs completed in under 12 minutes.  Today Python 2.7 takes 45 minutes to fail.

Look at [this](https://travis-ci.org/buildout/buildout/jobs/553658471): the slowest test (bootstrap.txt) used to take 14 seconds!  All other tests completed in under 10 seconds.

I'd say this is as much of a problem as the failures.

**Comment by reinout** [comment]
I added travis caching a few years ago. If I remember correctly: for caching the built-from-source python versions. Perhaps this caching mechanism has failed.

**Comment by mgedmin** [comment]
The slowdown is not in the CPython build part, but in the tests themselves.

[2 months ago](https://travis-ci.org/buildout/buildout/jobs/553658471):

```
 /home/travis/build/buildout/buildout/src/zc/buildout/downloadcache.txt (4.803 s)
 /home/travis/build/buildout/buildout/src/zc/buildout/dependencylinks.txt (6.281 s)
 /home/travis/build/buildout/buildout/src/zc/buildout/allowhosts.txt (3.248 s)
```

[Yesterday](https://travis-ci.org/buildout/buildout/jobs/574320042):

```
 /home/travis/build/buildout/buildout/src/zc/buildout/downloadcache.txt (74.492 s)
 /home/travis/build/buildout/buildout/src/zc/buildout/dependencylinks.txt (121.352 s)
 /home/travis/build/buildout/buildout/src/zc/buildout/allowhosts.txt (33.916 s)
```

How do 6 seconds become 120 seconds?  That's a 20x slowdown!

**Comment by mgedmin** [comment]
Here's me attempting to figure out what the tests are doing, by generating a flame graph with [py-spy](https://github.com/benfred/py-spy):

```
py-spy -d 30 -f /tmp/burn.svg -- bin/test -pvc -t allowhosts.txt
```

Fun fact!  GitHub does not support SVG images.  Here's a PNG:

![burn](https://user-images.githubusercontent.com/159967/63425808-13175200-c41a-11e9-851f-0de85ca0ead6.png)

You can see that about half of the time is spent waiting for subprocesses to finish running, and another half in handling HTTP requests (in select.select() that doesn't show up in the stack traces because it's a C function?).

**Comment by jamadden** [comment]
2 months ago, the build image was Ubuntu trusty (14.04) and took 4 minutes to boot. with `sudu: false` 

Yesterday, it was Ubuntu xenial (16.40) and took 10s to boot. 

Our configuration didn't change, but the travis defaults [did](https://changelog.travis-ci.com/xenial-as-the-default-build-environment-99476). I wonder what happens if we explicitly use `dist: trusty`?

**Comment by mgedmin** [comment]
The trusty  -> xenial switch is a very plausible reason for the breakage!

It'll be a bit more difficult to narrow down why this happened -- the build was building its own CPython from scratch, so any changes to OS-provided Python packages (or Python itself) shouldn't have any effect.

**Comment by reinout** [comment]
Running the tests locally takes a whopping 50 minutes (ubuntu 19.04), so it is not just travis that takes a lot of time. I also took a release from half a year ago and that also ran very slowly.

**Comment by mgedmin** [comment]
So I'm staring at the output of

    strace -tt -T -f -e trace=process -s 1024 -o /tmp/wat.trace bin/test -pvc -t allowhosts.txt

and scratching my head:

- 13:53:26.881056 to 13:53:40.125530 (~14s): things happen, including multiple os.popen('uname -p 2> /dev/null') calls
- 13:53:40.125530 to 13:53:45.992042 (~5s): there's a pause where no system calls at all are happening
- 13:53:45.992042 to 13:53:46.714009 (~1s): things happen
- 13:53:46.714009 to 13:53:52.516241 (~6s): no system calls at all
- 13:53:52.516241 to 13:53:52.895930 (~1s): things happen
- 13:53:52.895930 to 13:53:59.342853 (~7s): no system calls at all
- 13:53:59.342853 to 13:53:59.398473 (~0s): one final os.popen('uname -p ...')

And it took me this long to realize `-e trace=process` means I'm not seeing any socket I/O or filesystem operations, duh!

**Comment by mgedmin** [comment]
Hey actually this is a bit of a clue:

```
File "/home/mg/src/buildout/src/zc/buildout/allowhosts.txt", line 58, in allowhosts.txt
Failed example:
    print_(system(buildout), end='') # doctest: +ELLIPSIS
Expected:
    Develop: '/sample-buildout/allowdemo'
    Installing eggs...
    ...
    While:
      Installing eggs.
      Getting distribution for 'kss.core'.
    Error: Couldn't find a distribution for 'kss.core'.
Got:
    Getting distribution for 'setuptools'.
    Got setuptools 41.2.0.
    Upgraded:
      setuptools version 41.2.0;
    restarting.
    Generated script '/sample-buildout/bin/buildout'.
    Develop: '/sample-buildout/allowdemo'
    Installing eggs.
    Getting distribution for 'kss.core'.
    While:
      Installing eggs.
      Getting distribution for 'kss.core'.
    Error: Couldn't find a distribution for 'kss.core'.
```

Upgrading setuptools and restarting is going to take some time!  And this probably happens every time a doctest does a `system(buildout)`!

I generated a new trace with

    strace -tt -T -f -s 1024 -o /tmp/fulltrace.log bin/test -t allowhosts.txt

and grepped for the slowest system calls:

- the read() from a subprocess pipe in zc.buildout.testing in the main test runner process takes 9.4s to read a message saying `Getting distribution for 'setuptools`, when the child process does the upgrade
- another read() in the main process that sees `Develop: '/tmp/tmp4'`
- a child bin/buildout wrapper waited for python bin/buildout for 7.4s
- the parent /bin/sh of said bin/buildout wrapper waited for 17.3s
- another main process read() about "Develo" took 7s
- another /bin/sh wait() took 7s
- another main process read() about "Unused options for buildout: 'foo'" took 7.3s
- another /bin/sh wait() took 7.4s
- a select() in a thread of the main test runner process took 32s
- the full test run took 36s

A reminder of what this test does:

- do some setup that builds several eggs and sdists in a temp directory
- spawn a thread that responds to HTTP requests
- write some config files into a temp dir, run a bin/buildout
  - this buildout upgrades setuptools and re-execs itself
  - the re-exec'd child then runs a setup.py develop
- write some config files into a temp dir, run a bin/buildout
  - this buildout runs a setup.py develop
- write some config files into a temp dir, run a bin/buildout
  - this buildout doesn't run anything (other than uname -p which I ignore)

A regular child buildout run is 7.4s.  The one where it upgrades setuptools takes 17s, so 10s to upgrade then 7s to do the buildout things it does.

This explains part of the slowdown, but not all of it.  Remember, this test used to finish in 3 seconds total, which includes 3 subprocess runs.

I have one hypothesis:
- the test setup sets up an HTTP thread (about 4 seconds after process start)
- the HTTP thread simulates a package index
- the HTTP threads does some work and then sleeps for 32 seconds

Were the child processes supposed to use the local mock package index instead of the real PyPI?  Are they talking to the real PyPI, which happens to be much more slow?  (Yes they are if they can notice a new setuptools version and upgrade to it.)

**Comment by mgedmin** [comment]
Hm, the local HTTP server is used for the _setup_, but every test has to ask for it explicitly, and allowhosts.txt doesn't do that.

(Maybe the global PyPI was much faster 2 months ago?  Lol.)

**Comment by mgedmin** [comment]
> the test setup sets up an HTTP thread (about 4 seconds after process start)

4 seconds after process start seemed high to me, but it seems to be strace overhead.

py-spy captures show the doctest's setUp() gets executed in 1.4s, a whole 1s of which is spent inside easy_install_SetUp().

(There is a gratuitous time.sleep(0.25) inside zc.buildout.testing.wait(port, up) that shows up in the profile.)

**Comment by mgedmin** [comment]
Coming back to the original flamegraph (which was produced by a sampling profiler at 100 Hz and has lower overhead than strace), the three child bin/buildout invocations took
1. 15.81s
2. 5.98s
3. 5.74s

which is a bit less than the stace-shown 17s+7.4s+7.4s.

**Comment by reinout** [comment]
I've peppered the source code with print statements with `time.time()` output.
The line that takes a whole lot of time is in `easy_install.py`, `def _obtain(...)`. If you put a print statement before and after `if index.obtain(requirement) is None:`, you'll see it taking some 10+ seconds.

That index is something we subclass from setuptools. We only override the `url_ok()` method.
Add a `print()` with `time.time()` in there... You'll see almost 1000 calls to it. It starts like this:

    XXX 1566402062.6176834 _url_ok https://pypi.org/simple/setuptools/
    XXX 1566402062.7937553 _url_ok https://files.pythonhosted.org/packages/77/38/90cae8fd99c5dc402d25930e1c91f5fcf6de5c8e982fd152a5a061e3df56/setuptools-0.6b1-pyN.N.egg#sha256=ae0a6ec6090a92d08fe7f3dbf9f1b2ce889bce2a3d7724b62322a29b92cf93f0
    XXX 1566402062.7941427 _url_ok https://files.pythonhosted.org/packages/1f/ab/1ed6131b5112d1d47cb89e1cbea08f783f0ed49d6fc795ed2ff63f84d20c/setuptools-0.6b1-pyN.N.egg#sha256=377ff1ce6d8e24842fbc5f7fdf1c64875f7fe9a1c4db9e7b7479cbce1377da86
    XXX 1566402062.794522 _url_ok https://files.pythonhosted.org/packages/82/cb/906d051fdf1c8b7703d288d3cc44ff32746090378f7f8539c8476a21ce9a/setuptools-0.6b1.zip#sha256=24197f28b1e09f669a5d68eb9ff95f317ec9217f580c34fa73a4d7e6c03f81cc
    XXX 1566402062.7950497 _url_ok https://files.pythonhosted.org/packages/61/52/c84d22f3669ea8d4e54126f6241d7ed9828e3a382596f170043dfdaddbcc/setuptools-0.6b2-pyN.N.egg#sha256=eac0fde9b41f392bcb41ea04975ed41a2e52a067cc14a2737344d3105d26844c
    XXX 1566402062.7953436 _url_ok https://files.pythonhosted.org/packages/9b/87/48c155a10c4561fe9b8a83dcf9361ddb9cc9ab5c3ed1b504ac12ab11ed6a/setuptools-0.6b2-pyN.N.egg#sha256=34a05e133ae5bafc5a0a3b8ca6da29ea3b01a5919a6f0044292220ccf02fc32f

I don't know what's happening in that index.

The culprit we're looking for *might* be an update in setuptools instead of an update in ubuntu?

Anyway, I've got to go home now. I hope it helps :-)

**Comment by reinout** [comment]
Note: I ran it with `-t allowhosts.txt`. I modified the first call to buildout like this:

    >>> print_(system(buildout + ' -vvv'), end='') # doctest: +ELLIPSIS
    ddddDevelop: '/sample-buildout/allowdemo'
    Installing eggs...
    ...
    While:
      Installing eggs.
      Getting distribution for 'kss.core'.
    Error: Couldn't find a distribution for 'kss.core'.

And I stripped the rest from the file.

**Comment by icemac** [comment]
The TravisCI support enabled debug mode for this repository. It can be activated via an API call as described on https://docs.travis-ci.com/user/running-build-in-debug-mode/. This way we could inspect directly on the TravisCI server if needed.

**Comment by reinout** [comment]
I'd guess that such a slowdown ought to be visible in regular buildout-using projects.
I haven't actually worked in a small buildout-using project in the last few months. The zest.releaser tests take too long (with all its svn/git/hg calls) for a 10 second slowdown to be visible :-)

Does any of you have a not-too-big project with regular builds? It would be interesting if one of them suddenly started to take longer. Perhaps we can use it to figure out if anything changed in setuptools that interferes with buildout's hacks.

**Comment by icemac** [comment]
Most of the Zope projects are built using `buildout` and they have a history on TravisCI. But it is difficult to get actual and comparable time values because the container start-up time seems to be in the overall time displayed in the UI.

**Comment by icemac** [comment]
At least I found out, why the Python 2 tests are so fast and successful: They are not run because of a bug in `manuel`, see benji-york/Manuel#20.

**Comment by icemac** [comment]
I added a tracing feature to a fork of `manuel` to see which part of `buildout.txt` make TravisCI think it hangs. The result is, that it does not hang but the test needs > 10 min to complete and TravisCI kills the job after 10 min without writing anything. With this tracing feature enabled `buildout.txt` can be successfully run, see https://travis-ci.org/buildout/buildout/jobs/575714578

**Comment by gotcha** [comment]
I have started to run tests via Github Actions: see #494

#122 implies that tests setuptools 42.0.1 are now broken.

**Comment by jugmac00** [comment]
What is the current plan? Will Travis be fixed or will GitHub Actions be the future for running tests?

Buildout is vital for the Zope infrastructure and broken builds for more than 6 months do not feel good :-(

**Comment by idgserpro** [comment]
`setuptools >= 45` no longer works with Python 2. See:

https://github.com/pypa/setuptools/blob/a5dec2f14e3414e4ee5dd146bff9c289d573de9a/CHANGES.rst#v4500

Can `buildout` force `setuptools < 45` for Python 2, or is that not the responsibility of the `buildout`?

**Comment by gotcha** [comment]
When trying to fix Travis / run tests with GH actions, I bump into two issues:
- `setuptools.easy_install` command is deprecated since 42.0.0
- Python 2.7 deprecated since 45.0.0

And this is independent of the timeout issue.

I have started to speak to pypa team to understand what would be the right way to work around those two issues: https://github.com/pypa/setuptools/commit/b8101f06532b1deab448e6e23d0a61eb125c62df#r37539495

**Comment by idgserpro** [comment]
@gotcha the Python 2.7 problem is independent of your conversation with the `setuptools` people, do you agree? It may be necessary pin `setuptools` 44.0.0 in all tests with Python 2.7, do you agree? And this isn't going to be an easy job.

**Comment by gotcha** [comment]
@idgserpro Yes and no, the fix for Python 2.7 might depend on the fix needed to work around the `easy_install` deprecation. So I'd rather not make twice the not so easy job.

**Comment by idgserpro** [comment]
@gotcha the `setuptools` people have already said that they will not guarantee that new versions will work with Python 2. Can this situation change? It doesn't make sense to me test the buildout with a version of setuptools that is no longer supported.

**Comment by gotcha** [comment]
I managed to run the tests: see #502.
By chance, I found out that running the test suite in CentOS works much better.
I am looking for people with more expertise about innner Linux to explain what are the differences between CentOS and Ubuntu that justify the failures we see on Ubuntu.

That might help us to lessen the fragility of the test suite: it should not depend on a specific Linux distribution/knobs turned on or off.

**Comment by mgedmin** [comment]
Well, Ubuntu inherits a lot of Debian changes to Python packaging (splitting bits of the stdlib into separate installable deb packages), installation layouts (using dist-packages instead of site-packages), unvendoring of bits of pip that tend to cause problems when people start upgrading them piecewise, and such.

But I thought zc.buildout's tests built their own Pythons from sources specifically to avoid these kinds of inconsistencies?

**Comment by gotcha** [comment]
Right, the tests that were not passing on Ubuntu use their own Python built from source. And I kept it the same under CentOS

**Comment by gotcha** [comment]
Together with deep review, @mgedmin gently nudged for running tests under Debian.

The test suite was also very slow/hanging on Debian. I took the plunge and dove deep enough to find the culprit. The suite has a speed hack that depends on an environment variable. That environment was named `buildout-testing-index-url` which includes `-`. For a reason or another, this does not work on current Debian or Ubuntu.
I renamed the environment variable to `buildout_testing_index_url` and :tada: tests pass again on Travis see https://github.com/buildout/buildout/tree/2.x
